src/server/auth.rs

```diff
@@ -60,3 +64,88 @@ where
         token_type: TokenType::Agent,
     }
 }
+
+#[derive(Debug, Clone)]
+pub struct ACL {
+    cached_usernames: Arc<RwLock<HashSet<String>>>,
+    users: Vec<String>,
+    teams: Vec<(String, String)>,
+}
+
+impl ACL {
+    pub fn new(config: &Config, github: &GitHubApi) -> Result<Self> {
+        let mut users = Vec::new();
+        let mut teams = Vec::new();
+
+        for item in &config.server.bot_acl {
+            if let Some(middle) = item.find('/') {
+                let org = item[..middle].to_string();
+                let team = item[middle + 1..].to_string();
+                teams.push((org, team));
+            } else {
+                users.push(item.clone());
+            }
+        }
+
+        let acl = ACL {
+            cached_usernames: Arc::new(RwLock::new(HashSet::new())),
+            users,
+            teams,
+        };
+
+        acl.refresh_cache(github)?;
+        Ok(acl)
+    }
+
+    pub fn refresh_cache(&self, github: &GitHubApi) -> Result<()> {
+        // A new HashSet is created instead of clearing the old one
+        // This is done because if an error occurs the old cache is not flushed
+        let mut new_cache = HashSet::new();
+
+        for user in &self.users {
+            new_cache.insert(user.clone());
+        }
+
+        let mut orgs = HashMap::new();
+        for &(ref org, ref team) in &self.teams {
+            if let Err(err) = self.load_team(github, &mut new_cache, &mut orgs, org, team) {
+                warn!(
+                    "failed to authorize members of {}/{} to use the bot",
+                    org, team
+                );
+                warn!("caused by: {}", err);
+            }
+        }
+
+        // Update the shared cache
+        let mut cache = self.cached_usernames.write().unwrap();
+        *cache = new_cache;
+
+        Ok(())
+    }
+
+    fn load_team(
+        &self,
+        github: &GitHubApi,
+        new_cache: &mut HashSet<String>,
+        orgs: &mut HashMap<String, HashMap<String, usize>>,
+        org: &str,
+        team: &str,
+    ) -> Result<()> {
+        // Cache the list of teams in an org
+        if !orgs.contains_key(org) {
+            orgs.insert(org.to_string(), github.list_teams(org)?);
+        }
+
+        let members = github.team_members(orgs[org][team])?;
```

Comment:
If the team doesn't exist I think this will panic - maybe worth using `.get(team).ok_or("team doesn't exist")` or similar instead?

Reply:
Uh, woops. Fixed.